### DISCUSSION TITLE
Auxiliary losses

### DISCUSSION THREAD

**Discussion by EelcoHoogendoorn** [google/flax#2889] [Q&A]
I am trying to add an entropy regularizer to a probability distribution, and I am unsure of the preferred way of doing so in flax. I would know how to do it in a minimal example; but the hard part is how to do it in a composable manner, that will not results in all kinds of unwanted coupling between seperate framework components. That is, my distribution should go into a model which should go into a training env which goes into a recurrent state which goes into an RL training framework; and having to pass this loss term along all these intermediate steps feels like a lot of very undesirable coupling.

Ideally, I want some kind of distribution Module whos apply produces samples, but which then also has a mechanism for passing on a loss term. For weight based losses this is relatively simple since you can do it as a seperate step orthognoal to the model itself, via the optimizer. But any type of loss based on an intermediate activation makes things more compicated.

I have seen some examples involving the Module.sow method; though it seems very sparsely documented. But is this the way to go; and are there any examples of this being done in a nontrivial setting?

**Comment by EelcoHoogendoorn** [comment]
What ive elected to do thus far is adopt a convention of always returning an (output, loss) tuple from every apply method, and incorporate that into my RL framework as well, so i can collect these loss terms along a rollout. Its a bit coupled, but its doable and will allow me to add all kinds of activation based regularisation. 

Like I mentioned I bet this can be done with module.sow; but the lack of documentation, and lack of understanding on my part of how this might affect performance, including in pmapped settings where the pseudo-statefulness seems extra-scary, seems to weight against it in my mind; though im open to arguments to the contrary.

**Comment by cgarciae** [comment]
Hey @EelcoHoogendoorn, you can use collections to capture intermediate values, easiest way to do this is via the `sow` method:

```python

import jax
import jax.numpy as jnp
import flax.linen as nn

class Foo(nn.Module):
    def setup(self):
        self.dense = nn.Dense(10)

    def __call__(self, x):
        y = self.dense(x)

        # apply regularizer
        reg = jnp.sum(y ** 2)
        self.sow('losses', 'reg', reg)

        return y
    
foo = Foo()
variables = foo.init(jax.random.PRNGKey(0), jnp.ones((1, 10)))
params = variables['params']

y, updates = foo.apply({'params': params}, jnp.ones((1, 10)), mutable=['losses'])

print(updates['losses'])
```
```
FrozenDict({
    reg: (Array(16.529345, dtype=float32),),
})
```

**Comment by EelcoHoogendoorn** [comment]
Like I mentioned in my post, I am aware of Module.sow; I just dont understand it and its limitations given that I cannot find any documentation about its design intent.

**Reply by cgarciae** [reply]
Sorry about that, missed your reference to `sow`. 

> I just dont understand it and its limitations given that I cannot find any documentation about its design intent.

Its an simpler version of `Module.variable` mainly used when you just want to store intermediate values and collect them outside. The main limitation is that the collection in which you store the variables must be mutable, else it will behave like a no-op. All this is documented in [Module.sow](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.sow). If you have specific questions please send them here, we are working on improving our documentation.

**Reply by EelcoHoogendoorn** [reply]
Thanks, I appreciate the work done on flax a lot, and I realize documentation does not write itself.

Id love to know a little more about how sow and mutable collections work in general; it sure feels a bit 'black magical' in a functional context to have a mutable collection; itd be neat to know more of what goes into that sausage. Are there any tradeoffs in using mutable collections, either in functionality or performance, and so on?
